pi5/hw_pwm.go

```diff
@@ -166,6 +178,85 @@ func (pwm *pwmDevice) callGPIOReadall() {
 	}
 }
 
+/*
+For all Pins belonging to the same bank, pin data is stored contiguously and in 8 byte chunks.
+For a given pin, this method determines:
+ 1. which bank the pin belongs to
+ 2. the starting address of its 8 byte data chunk
+*/
+func getGPIOPinAddress(GPIONumber int) (int64, error) {
+
+	// Regarding the header: I (Maria) am unsure about what is stored here. It might just be GPIO 0.
+	// In that case, banks1 & 2 would be wrong for including the header in offset calcs.
+	bankHeaderSizeBytes := 8 // 8 bytes of either header data assosciated with a bank
+
+	if !(1 <= GPIONumber && GPIONumber <= maxGPIOPins) {
+		return -1, errors.New("pin is out of bank range")
+	}
+
+	// If we were using more than 1 bank, we'd have to subtract the GPIO Pin# from the first Pin of a given bank. See Maria's Guide to Pi5 Pincontrol for more info
+	pinBankOffset := GPIONumber
+	pinAddressOffset := bank0Offset + bankHeaderSizeBytes + ((pinBankOffset) * pinDataSize)
+	return int64(pinAddressOffset), nil
+}
+
+// This method updates the given mode of a pin by finding its specific location in memory & writing to the 'mode' byte in the 8 byte block of pin data.
+func (pwm *pwmDevice) writeToPinModeByte(GPIONumber int, newMode byte) error {
+
+	// Of the 8 bytes that represent a given pin's data, only the 4th index corresponds to the alternative mode setting
+	altModeIndex := 4
+
+	pinAddress, err := getGPIOPinAddress(GPIONumber)
+	if err != nil {
+		return fmt.Errorf("error getting gpio bank number: %w", err)
+	}
+
+	// find pin data within virtual page; retrieve the 4th byte from the pin data
+	vPage := *(pwm.gpioChipVPagePtr)
```

Comment:
you must pass a pointer to the virtual page. I tried without and it I think what happens is:

The virtual page is passed to the pwmDevice constructor before we set up the virtual page, so its completely empty . It also does not get updated as it changes because without the pointer we pass by value, not reference

Reply:
When you say must pass a pointer to the virtual page what do you mean? Line 215 is dereferencing the pointer to get the actual VPage object.